src/sentry/silo/client.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import annotations
+
+from typing import List
+
+from django.conf import settings
+
+from sentry.shared_integrations.client.base import BaseApiClient
+from sentry.silo.base import SiloMode
+from sentry.types.region import get_region_by_name
+
+DEFAULT_API_VERSION = "0"
+
+
+class SiloClientError(Exception):
+    """Indicates an error in processing a cross-silo HTTP request"""
+
+
+class BaseSiloClient(BaseApiClient):
+    integration_type = "silo_client"
+
+    @property
+    def access_modes() -> List[SiloMode]:
+        """
+        Limit access to the client to only the SiloModes set here.
+        """
+        raise NotImplementedError
+
+    def create_base_url(self, address: str) -> str:
+        api_prefix = f"/api/{self.api_version}"
```

Comment:
BIG OPEN QUESTION AHEAD, APOLOGIES: How does sentry's api versioning and schema upgrade policies look, in general?

Also this seems a bit odd though because it would imply that a client is locked into _every endpoint of a given version_.  How does adding a new, non backwards compatible endpoint work?  No client can use it until all other endpoints are also included in the new version namespace?

Reply:
To my knowledge, we haven't really had an upgrade policy since we often just issue new endpoints entirely, rather than version individual endpoints. You're correct that this would lock down any instantiated client to one API Version, but nothing is stopping someone from making multiple clients if they plan on using APIs at separate versions.

All the version does here is simplify the DevEx of providing a path to just whatever follows `/api/0/`, so we can totally remove it here and just have devs call `client.get('/api/0/organizations/')` vs `client.get('/organizations/)`, though I prefer to hide it away so that you cant end up making requests to non-api routes, e.g. "https://sentry.io/pricing" or something)

Reply:
I get that, so then my real question becomes: if this versioning schema is our own, we need to be thoughtful because it informs both ends of the protocol here.

Before I deploy a client, do I need to guarantee that all region silos also support that version number?   How do I as a developer ensure taht the region silos all consistently respond to X version?  Do I handle potential version mismatch (deployment drift) in my code?

If I'm renaming an attribute or upgrading its type in some way, is that a new api version?  Which schema or behavior upgrades are safe and which ones require a new version?  do old versions get deprecated?  when and how?

Are old versions of an api duplicated code?  do they share logic?  If they share logic, which changes to that shared logic are safe or not?

Who configures which version of an endpoint I should use?  Is that a settings value?  What, if any, guarantees, are made between two versions of an endpoint?  Should they be considered simply discrete endpoints, only historically related by name?

Reply:
These are all excellent call-outs that should definitely be nailed down a little bit more though it may not be worth while locking anything down in this PR specifically. For now, I think it's alright for us to assume `/api/0` as a prefix and for developers to assume the same version of Sentry will be using the client and receiving the request.

I'll add a some TODO-type functions to call this out, as well as the auth limitations, so we can revisit them later on.

Reply:
Sure!